kubernetes/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterScheduler.scala

```
@@ -70,121 +72,50 @@ private[spark] class KubernetesClusterScheduler(conf: SparkConf)
-                        }
-                    ]
-                },
-                {
-                    "name": "setup",
-                    "image": "$sparkDriverImage",
-                    "command": ["./install.sh"],
-                    "volumeMounts": [
-                        {
-                            "name": "workdir",
-                            "mountPath": "/work-dir"
-                        },
-                        {
-                            "name": "opt",
-                            "mountPath": "/opt"
-                        }
-                    ]
-                }
-            ]""")
-
-
     val labelMap = Map("type" -> "spark-driver")
     val pod = new PodBuilder()
       .withNewMetadata()
       .withLabels(labelMap.asJava)
       .withName(driverName)
-      .withAnnotations(annotationMap.asJava)
       .endMetadata()
       .withNewSpec()
       .withRestartPolicy("OnFailure")
+      .withServiceAccount(serviceAccountName)
       .addNewContainer()
       .withName("spark-driver")
       .withImage(sparkDriverImage)
       .withImagePullPolicy("Always")
-      .withCommand("/opt/spark/bin/spark-submit")
-      .withArgs(s"--class=${args.userClass}",
+      .withCommand(s"/opt/driver.sh")
+      .withArgs(s"$clientJarUri",
+                s"--class=${args.userClass}",
                 s"--master=$kubernetesHost",
                 s"--executor-memory=${driverDescription.mem}",
                 s"--conf=spark.executor.jar=$clientJarUri",
                 s"--conf=spark.executor.instances=$instances",
-                s"--conf=spark.kubernetes.namespace=${getNamespace()}",
+                s"--conf=spark.kubernetes.namespace=$nameSpace",
                 s"--conf=spark.kubernetes.driver.image=$sparkDriverImage",
-                s"--conf=spark.kubernetes.distribution.uri=$sparkDistUri",
-                "/work-dir/client.jar",
+                "/opt/spark/kubernetes/client.jar",
```

Why we need client.jar? If we assume Spark with k8s support is natively included we can also assume the Spark docker image should already have the classes we need right?

---

Good point. I think that is a healthy assumption to make. I believe this was a first attempt with room for improvement. Right now, this pr works by doing a curl on the provided url for the jar and saving it has `client.jar` and this is done in the `driver.sh` script.  I believe Taking a look at the scripts:  [driver.sh](https://github.com/erikerlandson/openshift-spark/blob/k8s-native/scripts/spark/added/driver.sh) and [executor.sh](https://github.com/erikerlandson/openshift-spark/blob/k8s-native/scripts/spark/added/executor.sh) from @erikerlandson repo will give you more clearer context.

---

Originally, we wrote that to use an [init container](http://kubernetes.io/docs/user-guide/production-pods/#handling-initialization) to download the `client.jar` into each executor and driver pod. I think this approach is fine, because init-containers are relatively new and this would work with older releases of k8s as well. @tnachen, some users bake the client-jar into the docker image they use, and some would provide a network url and expect it to get downloaded. We'd need a way to support both of these I imagine.

---

My opinion on this is:
(a) spark should be assumed to be installed on docker images
(b) the user's application jar should NOT be assumed to be on images.
The user's application jar is a "volatile" - it can change with every submission.   I would not want to be re-creating docker images every time I wanted to submit an app.  But I absolutely want to be using docker images with spark pre-installed.

---

To elaborate, I don't think it would be very hard to support an image with user app jars pre-installed, nor would it be hard to download a spark-distro.jar and unpack it with annotations, it just seems non-idiomatic to me.   (And the annotations support is currently wonky.   My openshift install didn't recognize it, and _failed silently_)

---

I see, I thought this is Spark itself and from you describe it's the user application's jar.
We will definitely need to fix this if we need to support python or R, but for now this is fine.
And I'll also rename this to application or app jar just to stay within the Spark nomenclature.

---

To support images with user client.jar preinstalled, we could make path-to-client-jar configurable, as the location for downloading to, or alternatively the location to assume it already lives in the image

---

That sounds good to me.

---

good point, "jar" should be generalized to include python/R counterparts

---

One other consideration about location of application jar file (or python, R, etc):  When pods are running in non-admin mode, it matters where that file goes, or more specifically that the container user (which can be non-root) has write privileges, if it is downloading.  For example, if it downloads to `/opt/` then `/opt` needs to be writable by non-root or the download will fail.  It wasn't a problem when submitting against default kubernetes but it came up immediately when I tested against my openshift cluster, logged in as 'developer'